flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutorGateway.java

```diff
@@ -102,12 +102,12 @@
 		@RpcTimeout Time timeout);
 
 	/**
-	 * Batch release intermediate result partitions.
-	 *
+	 * Batch release/promote intermediate result partitions.
 	 * @param jobId id of the job that the partitions belong to
-	 * @param partitionIds partition ids to release
+	 * @param partitionToRelease partition ids to release
+	 * @param partitionsToPromote partitions ids to promote
 	 */
-	void releasePartitions(JobID jobId, Collection<ResultPartitionID> partitionIds);
+	void releaseOrPromotePartitions(JobID jobId, Collection<ResultPartitionID> partitionToRelease, Collection<ResultPartitionID> partitionsToPromote);
```

Comment:
Do we strictly need both `partitionsToRelease` and `partitionsToPromote`? One could also say that one of them is the difference from the set of registered partitions and the specified argument. E.g. `partitionsToPromote` could be everything which is kept in the `PartitionTable` after `partitionsToRelease` has been removed.

I think it could decrease a bit the complexity at the cost of explicitness because it would avoid the situation where a partition is in neither of the specified sets `partitionToRelease` and `partitionsToPromote`.

Reply:
I don't think that would work since this method is called more than once.

If it were only called at the end the job, then yes the promotion argument is redundant. However, we also call this method in cases where pipelined regions are reset; in this case you would implicitly promote partitions of preceding pipelined regions.

Reply:
Although...I suppose if the promotion were done at the time the JM disconnects this could work.

Reply:
But then we'd have to differentiate properly between (un)expected disconnects as otherwise the safety net no longer works.

Reply:
I'm wondering though, are there any guarantees that the disconnect is processed after all messages have arrived? Otherwise it might be that a release call is missed and partitions are promoted although they shouldn't be.

This would be a bigger issue than a partition not being in either set, as in this case the partition will be cleaned up eventually when the JM exits.

Reply:
If there is a case where we want to release a sub set of the partitions while keeping other w/o promoting them to cluster partitions, then I think this extension is justified.